### PR TITLE
Datatx pull model

### PR DESCRIPTION
**Pull request by redblom**
This PR implements CS3 tx (pull model) (https://github.com/cs3org/cs3apis/pull/143) using rclone as data transfer tool.
The following commands are available: `transfer-create`, `transfer-list`, `transfer-get-status`, `transfer-retry`, `transfer-cancel`

Create a transfer (sender): 
```
#  ./cmd/reva/reva transfer-create -grantee 3000989c-6gh5-4cec-a9ce-e876tf7de452 -granteeType user -idp cesnet.cz /home/data/big-data
+--------------------------------------+-------------+--------------------------------------+-----------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------+-------------+--------------------------------------+---------------------+-------------------------------+-------------------------------+
| #                                    | OWNER.IDP   | OWNER.OPAQUEID                       | RESOURCEID                                                                                          | PERMISSIONS                                                                                                                                                                                                     | TYPE              | GRANTEE.IDP | GRANTEE.OPAQUEID                     | SHARETYPE           | CREATED                       | UPDATED                       |
+--------------------------------------+-------------+--------------------------------------+-----------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------+-------------+--------------------------------------+---------------------+-------------------------------+-------------------------------+
| 775b47bd-4cca-4470-86f7-0424b7a998ab | surfsara.nl | 4029579c-6ad5-4cec-a9ce-e843f77de452 | storage_id:"123e4567-e89b-12d3-a456-426655440000" opaque_id:"fileid-jimmie%2Fdata%2Fdata/big-data"  | permissions:<create_container:true delete:true get_path:true initiate_file_download:true initiate_file_upload:true list_container:true list_file_versions:true move:true restore_file_version:true stat:true >  | GRANTEE_TYPE_USER | cesnet.cz   | 3000989c-6gh5-4cec-a9ce-e876tf7de452 | SHARE_TYPE_TRANSFER | 2021-09-06 10:52:47 +0000 UTC | 2021-09-06 10:52:47 +0000 UTC |
+--------------------------------------+-------------+--------------------------------------+-----------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------+-------------+--------------------------------------+---------------------+-------------------------------+-------------------------------+
```
Discover and accept the transfer (receiver):
```
# ./cmd/reva/reva ocm-share-list-received
+--------------------------------------+-------------+--------------------------------------+----------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------+-------------+--------------------------------------+-------------------------------+-------------------------------+----------------------+---------------------+
| #                                    | OWNER.IDP   | OWNER.OPAQUEID                       | RESOURCEID                                                                                               | PERMISSIONS                                                                                                                                                       | TYPE              | GRANTEE.IDP | GRANTEE.OPAQUEID                     | CREATED                       | UPDATED                       | STATE                | SHARETYPE           |
+--------------------------------------+-------------+--------------------------------------+----------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------+-------------+--------------------------------------+-------------------------------+-------------------------------+----------------------+---------------------+
| 40ff72c6-c30c-4380-827d-5e2d3970b57c | surfsara.nl | 4029579c-6ad5-4cec-a9ce-e843f77de452 | storage_id:"123e4567-e89b-12d3-a456-426655440000" opaque_id:"fileid-jimmie%2Fdata%2Fbig-data"       | permissions:<get_path:true get_quota:true initiate_file_download:true list_grants:true list_container:true list_file_versions:true list_recycle:true stat:true >  | GRANTEE_TYPE_USER | cesnet.cz   | 3000989c-6gh5-4cec-a9ce-e876tf7de452 | 2021-09-06 10:52:47 +0000 UTC | 2021-09-06 10:52:47 +0000 UTC | SHARE_STATE_PENDING  | SHARE_TYPE_TRANSFER |
+--------------------------------------+-------------+--------------------------------------+----------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------+-------------+--------------------------------------+-------------------------------+-------------------------------+----------------------+---------------------+

# ./cmd/reva/reva ocm-share-update-received -state accepted 40ff72c6-c30c-4380-827d-5e2d3970b57c
OK
```
Receiver discovers the transfer ID needed to get the status of, retry, or cancel the transfer:
```
# ./cmd/reva/reva transfer-list -shareId 40ff72c6-c30c-4380-827d-5e2d3970b57c
+--------------------------------------+--------------------------------------+
| SHAREID.OPAQUEID                     | ID.OPAQUEID                          |
+--------------------------------------+--------------------------------------+
| 40ff72c6-c30c-4380-827d-5e2d3970b57c | 76508e0e-eda1-43fc-9b4d-c6b526b5a75c |
+--------------------------------------+--------------------------------------+
```
Get the transfer status:
```
# ./cmd/reva/reva transfer-get-status -txId 76508e0e-eda1-43fc-9b4d-c6b526b5a75c
+--------------------------------------+--------------------------------------+----------------+-----------------------------------+
| SHAREID.OPAQUEID                     | ID.OPAQUEID                          |         STATUS | CTIME                             |
+--------------------------------------+--------------------------------------+----------------+-----------------------------------+
| 40ff72c6-c30c-4380-827d-5e2d3970b57c | 76508e0e-eda1-43fc-9b4d-c6b526b5a75c | STATUS_TRANSFER_COMPLETE | Mon Sep 6 10:53:08 +0000 UTC 2021 |
+--------------------------------------+--------------------------------------+----------------+-----------------------------------+
```
Retry a failed transfer:
```
# ./cmd/reva/reva transfer-retry -txId 76508e0e-eda1-43fc-9b4d-c6b526b5a75c
+--------------------------------------+--------------------------------------+---------------------+-------------------------------+
| SHAREID.OPAQUEID                     | ID.OPAQUEID                          |              STATUS | CTIME                         |
+--------------------------------------+--------------------------------------+---------------------+-------------------------------+
| 40ff72c6-c30c-4380-827d-5e2d3970b57c | 76508e0e-eda1-43fc-9b4d-c6b526b5a75c | STATUS_TRANSFER_NEW | 2021-09-06 10:53:08 +0000 UTC |
+--------------------------------------+--------------------------------------+---------------------+-------------------------------+
```
Cancel a running transfer:
```
# ./cmd/reva/reva transfer-cancel -txId f82396c4-386b-484f-856c-b75936980f1e
+--------------------------------------+--------------------------------------+---------------------------+-----------------------------------+
| SHAREID.OPAQUEID                     | ID.OPAQUEID                          |                    STATUS | CTIME                             |
+--------------------------------------+--------------------------------------+---------------------------+-----------------------------------+
| a940df85-9023-4cf8-809a-4c33748f2c66 | f82396c4-386b-484f-856c-b75936980f1e | STATUS_TRANSFER_CANCELLED | Wed Sep 1 11:56:25 +0000 UTC 2021 |
+--------------------------------------+--------------------------------------+---------------------------+-----------------------------------+
```

### CODE REVIEW

**Comment by redblom** [comment]
@labkode @ishank011 please review

File: internal/grpc/services/datatx/datatx.go [hunk 1]
```diff
@@ -81,20 +156,228 @@ func (s *service) UnprotectedEndpoints() []string {
...
+	}
+	s.txShareDriver.Lock()
+	defer s.txShareDriver.Unlock()
+
+	s.txShareDriver.model.TxShares[txInfo.GetId().OpaqueId] = txShare
+	if err := s.txShareDriver.model.saveTxShare(); err != nil {
+		err = errors.Wrap(err, "datatx service: error saving transfer share: "+datatx.Status_STATUS_INVALID.String())
+		return &datatx.PullTransferResponse{
+			Status: status.NewInvalid(ctx, "error pulling transfer"),
+		}, err
```
Comment (redblom) [inline internal/grpc/services/datatx/datatx.go:194]: @ishank011 : I wonder, what are the criteria for returning the error vs. setting the response status error ?

**Comment by labkode** [comment]
@redblom the PR is ready, can you just rebase on top of master and then we merge it with the cs3apis changes?

**Comment by redblom** [comment]
> @redblom the PR is ready, can you just rebase on top of master and then we merge it with the cs3apis changes?

@labkode : Done!

**Review by glpatcern — DISMISSED** [review]
As this work was already reviewed and it was simply pending the CI, I'm approving it.

**Review by ishank011 — APPROVED** [review]
One comment and then we can merge

**Review by ishank011 — APPROVED** [review]

**Pull request merged**